### ISSUE TITLE
Tuning pegasos parameters

### ISSUE THREAD

**Issue by mrgloom** [davisking/dlib#49]
I'm trying to tune pegasos parameters without no luck.

Here is simple example using 6 one-dimensional points. 

~~~
void dlib_svm_test()
{
	for(double lambda= 10e-10;lambda<=10e10;lambda*=10)
	{
		typedef dlib::matrix<double, 0, 1> sample_type;
		typedef dlib::linear_kernel<sample_type> kernel_type;

		dlib::svm_pegasos<kernel_type> pegasos;

		cout << lambda <<endl;

		pegasos.set_lambda(lambda);
		pegasos.set_kernel(kernel_type());

		std::vector<sample_type> samples(6);
		sample_type sample;
		sample.set_size(1);

		sample(0)= 188.0d;
		samples[0]= sample;
		sample(0)= 168.0d;
		samples[1]= sample;
		sample(0)= 191.0d;
		samples[2]= sample;
		sample(0)= 150.0d;
		samples[3]= sample;
		sample(0)= 154.0d;
		samples[4]= sample;
		sample(0)= 124.0d;
		samples[5]= sample;

		pegasos.train(samples[0],+1);
		pegasos.train(samples[1],+1);
		pegasos.train(samples[2],+1);
		pegasos.train(samples[3],-1);
		pegasos.train(samples[4],-1);
		pegasos.train(samples[5],-1);

		cout<< pegasos(samples[0]) <<endl;
		cout<< pegasos(samples[1]) <<endl;
		cout<< pegasos(samples[2]) <<endl;
		cout<< pegasos(samples[3]) <<endl;
		cout<< pegasos(samples[4]) <<endl;
		cout<< pegasos(samples[5]) <<endl;

		pegasos.clear();
	}
}
~~~

The output tha tI get:
~~~
0.0000000010
-3963387.1199921928
-3541750.1923335334
-4026632.6591409920
-3162276.9574407390
-3246604.3429724714
-2614148.9514844813
0.0000000100
-1253333.0548153266
-1119999.7511116527
-1273333.0503708781
-999999.7777783460
-1026666.4385190808
-826666.4829635697
0.0000001000
-396338.7119995961
-354175.0192337657
-402663.2659144707
-316227.6957445183
-324660.4342976844
-261414.8951489388
0.0000010000
-125333.3054819095
-111999.9751115777
-127333.3050374593
-99999.9777782790
-102666.6438523454
-82666.6482968476
0.0000100000
-39633.8712003365
-35417.5019237890
-40266.3265918186
-31622.7695748963
-32466.0434302058
-26141.4895153846
0.0001000000
-12533.3305485679
-11199.9975115703
-12733.3305041176
-9999.9977782724
-10266.6643856720
-8266.6648301755
0.0010000000
-3963.3871204108
-3541.7501927916
-4026.6326595536
-3162.2769579343
-3246.6043434582
-2614.1489520294
0.0100000000
-1253.3330552344
-1119.9997515702
-1273.3330507840
-999.9997782725
-1026.6664390053
-826.6664835091
0.1000000000
-396.3387124203
-354.1750196940
-402.6632663292
-316.2276962404
-324.6604347856
-261.4148956963
1.0000000000
-125.3333059077
-111.9999755772
-127.3333054573
-99.9999782797
-102.6666443458
-82.6666488500
10.0000000000
-39.6338716427
-35.4175024067
-40.2663270281
-31.6227700943
-32.4660439415
-26.1414900875
100.0000000000
-12.5333310483
-11.1999980544
-12.7333309973
-9.9999983600
-10.2666649587
-8.2666654680
1000.0000000000
-3.7091542406
-3.3145634810
-3.7683428546
-2.9594317974
-3.0383499493
-2.4464638100
10000.0000000000
-0.4292670207
-0.3836003494
-0.4361170215
-0.3425003451
-0.3516336794
-0.2831336723
100000.0000000000
0.0372866667
0.0333200000
0.0378816667
0.0297500000
0.0305433333
0.0245933333
1000000.0000000000
0.0037286667
0.0033320000
0.0037881667
0.0029750000
0.0030543333
0.0024593333
10000000.0000000000
0.0003728667
0.0003332000
0.0003788167
0.0002975000
0.0003054333
0.0002459333
100000000.0000000000
0.0000372867
0.0000333200
0.0000378817
0.0000297500
0.0000305433
0.0000245933
1000000000.0000000000
0.0000037287
0.0000033320
0.0000037882
0.0000029750
0.0000030543
0.0000024593
10000000000.0000000000
0.0000003729
0.0000003332
0.0000003788
0.0000002975
0.0000003054
0.0000002459
100000000000.0000000000
0.0000000373
0.0000000333
0.0000000379
0.0000000297
0.0000000305
0.0000000246
~~~

So the problem that I get all samples predicted as negative or positive.

**Comment by davisking** [comment]
This is an online learning algorithm based on stochastic gradient descent.  Each time you call train() it takes one gradient step, so you must call train() way more than 6 times.  

You are also probably better of using a batch algorithm rather than an online algorithm.  Use this guide to pick one that is appropriate for your task: http://dlib.net/ml_guide.svg

**Comment by mrgloom** [comment]
This is just example and in real application I have lots of data that can't fit in memory and I need online learning algorithm.

Tried several iterations through data, no result.

~~~
    typedef dlib::matrix<double, 0, 1> sample_type;
    typedef dlib::linear_kernel<sample_type> kernel_type;

    dlib::svm_pegasos<kernel_type> pegasos;
	
    for(double lambda= 10e-10;lambda<=10e10;lambda*=10)
    {
        cout << lambda <<endl;

        pegasos.set_lambda(lambda);
        pegasos.set_kernel(kernel_type());

        std::vector<sample_type> samples(6);
        sample_type sample;
        sample.set_size(1);

        sample(0)= 188.0d;
        samples[0]= sample;
        sample(0)= 168.0d;
        samples[1]= sample;
        sample(0)= 191.0d;
        samples[2]= sample;
        sample(0)= 150.0d;
        samples[3]= sample;
        sample(0)= 154.0d;
        samples[4]= sample;
        sample(0)= 124.0d;
        samples[5]= sample;

		for(int iter=0;iter<1000;++iter)    
		{
        	pegasos.train(samples[0],+1);
        	pegasos.train(samples[1],+1);
        	pegasos.train(samples[2],+1);
        	pegasos.train(samples[3],-1);
        	pegasos.train(samples[4],-1);
        	pegasos.train(samples[5],-1);

			if(pegasos(samples[0])==+1 && pegasos(samples[1])==+1 && pegasos(samples[2])==+1 &&
			pegasos(samples[3])==-1 && pegasos(samples[4])==-1 && pegasos(samples[5])==-1 )
			{
				cout<<"OK."<<endl;
			}
    	}

		pegasos.clear();
	}
	cout<<"done!"<<endl;
}
~~~

**Comment by mrgloom** [comment]
Maybe this can help: http://stackoverflow.com/questions/33174696/tuning-vlfeat-svm/   , but it seems pegasos have different SVM formulation and small learning rate doesn't help?

**Comment by davisking** [comment]
The output isn't going to be +1 or -1.  It's some number that is > 0.

**Comment by mrgloom** [comment]
Yes,  my mistake.

Here is new code:
~~~
void dlib_svm_test3()
{
    typedef dlib::matrix<double, 0, 1> sample_type;
    typedef dlib::linear_kernel<sample_type> kernel_type;

    dlib::svm_pegasos<kernel_type> pegasos;

    int n_iter= 10000;
    for(double lambda= 10e-10;lambda<=10e10;lambda*=10)
    {
        cout << lambda <<endl;

        pegasos.set_lambda(lambda);
        pegasos.set_kernel(kernel_type());

        std::vector<sample_type> samples(6);
        sample_type sample;
        sample.set_size(1);

        sample(0)= 188.0d;
        samples[0]= sample;
        sample(0)= 168.0d;
        samples[1]= sample;
        sample(0)= 191.0d;
        samples[2]= sample;
        sample(0)= 150.0d;
        samples[3]= sample;
        sample(0)= 154.0d;
        samples[4]= sample;
        sample(0)= 124.0d;
        samples[5]= sample;

        for(int iter=0;iter<n_iter;++iter)
        {
            pegasos.train(samples[0],+1);
            pegasos.train(samples[1],+1);
            pegasos.train(samples[2],+1);
            pegasos.train(samples[3],-1);
            pegasos.train(samples[4],-1);
            pegasos.train(samples[5],-1);

            if(		sign(pegasos(samples[0]))==+1 &&
            		sign(pegasos(samples[1]))==+1 &&
					sign(pegasos(samples[2]))==+1 &&
            		sign(pegasos(samples[3]))==-1 &&
					sign(pegasos(samples[4]))==-1 &&
					sign(pegasos(samples[5]))==-1 )
            {
                cout<<"OK."<<endl;
            }
        }

        cout << pegasos(samples[0]) << endl;
        cout << pegasos(samples[1])	<< endl;
		cout << pegasos(samples[2]) << endl;
        cout << pegasos(samples[3]) << endl;
        cout << pegasos(samples[4]) << endl;
		cout << pegasos(samples[5]) << endl;

        pegasos.clear();
    }
    cout<<"done!"<<endl;
}
~~~
Here is output:
~~~
0.0000000010
-5944882.5041987421
-5312448.1981508601
-6039747.6501059243
-4743257.3227077676
-4869744.1839173436
-3921092.7248455212
0.0000000100
-1879936.9099174207
-1679943.6247028753
-1909935.9026996025
-1499949.6680097843
-1539948.3250526933
-1239958.3972308752
0.0000001000
-594488.2487954451
-531244.8217441718
-603974.7628531362
-474325.7373980258
-486974.4228082804
-392109.2822313703
0.0000010000
-187993.7106111174
-167994.3865077624
-190993.6092266206
-149994.9948147430
-153994.8596354139
-123995.8734803815
0.0000100000
-44927.9577496871
-40148.7209681511
-45644.8432669175
-35847.4078647686
-36803.2552210758
-29634.4000487717
0.0001000000
-2712.3218170722
-2424.0120868655
-2755.5682766032
-2164.5333296795
-2222.1952757208
-1789.7306804107
0.0010000000
-450.6151568366
-402.7038160088
-457.8018579607
-359.5836092637
-369.1658774293
-297.2988661876
0.0100000000
-18.6067623428
-16.6300149115
-18.9032744575
-14.8509422233
-15.2462917096
-12.2811705626
0.1000000000
-2.3438240832
-2.0947721142
-2.3811818786
-1.8706253420
-1.9204357358
-1.5468577823
1.0000000000
-0.0352592048
-0.0315925381
-0.0358092048
-0.0282925381
-0.0290258714
-0.0235258714
10.0000000000
0.8769121278
0.7836137410
0.8909068858
0.6996451928
0.7183048702
0.5783572900
100.0000000000
0.9751781577
0.8714348244
0.9907396577
0.7780658245
0.7988144911
0.6431994912
1000.0000000000
0.9831002410
0.8785150335
0.9987880222
0.7843883466
0.8053053881
0.6484275768
10000.0000000000
0.3727864533
0.3331283200
0.3787351733
0.2974360000
0.3053676266
0.2458804266
100000.0000000000
0.0372866667
0.0333200000
0.0378816667
0.0297500000
0.0305433333
0.0245933333
1000000.0000000000
0.0037286667
0.0033320000
0.0037881667
0.0029750000
0.0030543333
0.0024593333
10000000.0000000000
0.0003728667
0.0003332000
0.0003788167
0.0002975000
0.0003054333
0.0002459333
100000000.0000000000
0.0000372867
0.0000333200
0.0000378817
0.0000297500
0.0000305433
0.0000245933
1000000000.0000000000
0.0000037287
0.0000033320
0.0000037882
0.0000029750
0.0000030543
0.0000024593
10000000000.0000000000
0.0000003729
0.0000003332
0.0000003788
0.0000002975
0.0000003054
0.0000002459
100000000000.0000000000
0.0000000373
0.0000000333
0.0000000379
0.0000000297
0.0000000305
0.0000000246
~~~

Seems it cross zero at some point near 1-10 and I narrow down lambda interval:
~~~
for(double lambda= 1;lambda<=10;lambda+= 0.5)
~~~
here is output:
~~~
1.0000000000
-0.0352592048
-0.0315925381
-0.0358092048
-0.0282925381
-0.0290258714
-0.0235258714
 
1.5000000000
0.3995472902
0.3569822683
0.4059320435
0.3186737486
0.3271867529
0.2633392201
 
2.0000000000
0.5622313194
0.5023731064
0.5712100514
0.4485007146
0.4604723572
0.3706850377
 
2.5000000000
0.5666162407
0.5063001785
0.5756636500
0.4520157225
0.4640789349
0.3736048417
 
3.0000000000
0.6499896349
0.5808100101
0.6603665787
0.5185483478
0.5323842728
0.4286148355
 
3.5000000000
0.6450256068
0.5763784195
0.6553226849
0.5145959510
0.5283253884
0.4253546075
 
4.0000000000
0.7126063061
0.6367729727
0.7239813061
0.5685229728
0.5836896394
0.4699396394
 
4.5000000000
0.8039196086
0.7183746912
0.8167513462
0.6413842655
0.6584932490
0.5301758729
 
5.0000000000
0.7786487433
0.6957943064
0.7910769088
0.6212253131
0.6377962005
0.5135145451
 
5.5000000000
0.7752642402
0.6927715779
0.7876381396
0.6185281819
0.6350267143
0.5112877209
 
6.0000000000
0.7860099861
0.7023756043
0.7985551434
0.6271046606
0.6438315369
0.5183799641
 
6.5000000000
0.7979038805
0.7130054141
0.8106386504
0.6365967943
0.6535764876
0.5262287880
 
7.0000000000
0.8299289130
0.7416245878
0.8431745618
0.6621506951
0.6798115601
0.5473550723
 
7.5000000000
0.8682796807
0.7758964050
0.8821371720
0.6927514569
0.7112281120
0.5726531985
 
8.0000000000
0.8374884918
0.7483816797
0.8508545137
0.6681855488
0.6860069112
0.5523466930
 
8.5000000000
0.8283264739
0.7401950552
0.8415461867
0.6608767784
0.6785030621
0.5463059341
 
9.0000000000
0.8626082275
0.7708304497
0.8763748941
0.6882304498
0.7065860053
0.5689193387
 
9.5000000000
0.9013222806
0.8054265563
0.9157066392
0.7191204044
0.7382995493
0.5944559629
 
10.0000000000
0.8769121278
0.7836137410
0.8909068858
0.6996451928
0.7183048702
0.5783572900
~~~

It's strange but I get exact answer only 2 times(maybe it's because stochastic nature of algorithm?):
~~~
OK.
lambda: 2.0000000000
iter: 4908
OK.
lambda: 2.0000000000
iter: 5099
~~~

Any ideas on what is may be wrong I how to debug it? Or only way is to iterate over the large interval of lambda with small step size and fixed number of iterations over the dataset?

One more question, did you tried to replace linear SVM with Pegasos in Linear SVM + HOG setting for object detection? I think with "online learning" we can remove step of hard negative mining and just iterate over data.

**Comment by davisking** [comment]
It's not a bug.  This is just how the pegasos algorithm is, which I agree
is not very user friendly.  So I would suggest using something else if it's
not working out for you.

**Comment by mrgloom** [comment]
Hello, also I wonder does Pegasos algorithm need input data to be shuffled?
For example this sequence will be bad `-1-1-1-1-1-1-1-1-1+1+1+1+1+1+1+1+1` and this is good `+1-1+1-1+1-1+1-1+1+1` for training?

Also can we train any other SVM algorithm with linear kernel implemented in Dlib on all available data and then "finetune" weights on new data in "online learning" way?

**Comment by davisking** [comment]
Yes, the data should be shuffled.  You can do a variety of kinds of "fine
tuning".  However, for linear SVMs it's almost always better to train them
in batch mode with all the data.  For example, the svm_c_linear_dcd_trainer
is fundamentally superior to pegasos for training linear SVMs.  It is both
faster and more accurate.

**Comment by mrgloom** [comment]
Can you clarify more on this? [Data should be shuffled because of decay of learning rate](http://stackoverflow.com/questions/29516033/stochastic-gradient-descent-for-regression-with-shuffle-data-performs-better-tha
)? What if learning rate is constant? How learning rate decay in pegasos dlib implementation?

What if I train pegasos algorithm using shuffled data with negative and positive examples and then I want to "finetune" it using only unseen negative examples?
